File: ailee/domains/video_temporal_provenance/governor.py
# Copyright (c) Don Michael Feeney Jr.
# Licensed under the MIT License.
"""
Governor for Video Temporal Provenance domain.
"""

from dataclasses import dataclass
from typing import List, Optional
from .config import VideoTemporalConfig
from .policy import VideoTemporalPolicy
from .ffi import TPEFFIWrapper, TemporalIntegrityMetricsPy

@dataclass
class FrameSignal:
    frame_index: int
    timestamp_sec: float
    raw_trust: float
    hash_delta: float
    dx: float
    dy: float
    flow_consistency: float

@dataclass
class VideoGovernorDecision:
    overall_trust_score: float
    safety_status: str
    anomaly_count: int
    reason: str
    metrics: TemporalIntegrityMetricsPy

class VideoTemporalGovernor:
    def __init__(self, config: Optional[VideoTemporalConfig] = None, policy: Optional[VideoTemporalPolicy] = None):
        self.config = config or VideoTemporalConfig()
        self.policy = policy or VideoTemporalPolicy()
        self.wrapper = TPEFFIWrapper()

    def reset(self):
        self.wrapper.reset()

    def evaluate_sequence(self, frames: List[FrameSignal]) -> VideoGovernorDecision:
        self.wrapper.reset()
        for f in frames:
            self.wrapper.ingest_frame(
                f.frame_index,
                f.timestamp_sec,
                f.raw_trust,
                f.hash_delta,
                f.dx,
                f.dy,
                f.flow_consistency
            )

        metrics = self.wrapper.evaluate()

        status = metrics.safety_status
        reason = "Sequence evaluated successfully."

        if metrics.overall_trust_score < self.config.min_trust_threshold:
            status = "OUTRIGHT_REJECTED"
            reason = f"Overall trust {metrics.overall_trust_score:.1f} below minimum threshold {self.config.min_trust_threshold:.1f}."
        elif metrics.anomaly_count > self.config.max_allowed_anomalies:
            status = "PARTIALLY_TRUSTED" if status != "OUTRIGHT_REJECTED" else status
            reason = f"Anomaly count {metrics.anomaly_count} exceeded max allowed {self.config.max_allowed_anomalies}."

        return VideoGovernorDecision(
            overall_trust_score=metrics.overall_trust_score,
            safety_status=status,
            anomaly_count=metrics.anomaly_count,
            reason=reason,
            metrics=metrics
        )

File: packages/ailee-ts/src/domains/video_temporal_provenance/governor.ts
//! Copyright (c) Don Michael Feeney Jr.
//! Licensed under the MIT License.

import { VideoTemporalConfig, DEFAULT_VIDEO_TEMPORAL_CONFIG } from "./config.js";
import { VideoTemporalPolicy, DEFAULT_VIDEO_TEMPORAL_POLICY } from "./policy.js";
import { FrameSignal, TemporalIntegrityMetrics, TPEFFIBridge } from "./ffi.js";

export interface VideoGovernorDecision {
  overallTrustScore: number;
  safetyStatus: "ACCEPTED" | "PARTIALLY_TRUSTED" | "OUTRIGHT_REJECTED";
  anomalyCount: number;
  reason: string;
  metrics: TemporalIntegrityMetrics;
}

export class VideoTemporalGovernor {
  private config: VideoTemporalConfig;
  private policy: VideoTemporalPolicy;
  private bridge: TPEFFIBridge;

  constructor(
    config: VideoTemporalConfig = DEFAULT_VIDEO_TEMPORAL_CONFIG,
    policy: VideoTemporalPolicy = DEFAULT_VIDEO_TEMPORAL_POLICY
  ) {
    this.config = config;
    this.policy = policy;
    this.bridge = new TPEFFIBridge();
  }

  public reset(): void {
    this.bridge.reset();
  }

  public evaluateSequence(frames: FrameSignal[]): VideoGovernorDecision {
    this.bridge.reset();
    for (const f of frames) {
      this.bridge.ingestFrame(f);
    }

    const metrics = this.bridge.evaluate();
    let status = metrics.safetyStatus;
    let reason = "Sequence evaluated successfully.";

    if (metrics.overallTrustScore < this.config.minTrustThreshold) {
      status = "OUTRIGHT_REJECTED";
      reason = `Overall trust ${metrics.overallTrustScore.toFixed(1)} below threshold ${this.config.minTrustThreshold.toFixed(1)}.`;
    } else if (metrics.anomalyCount > this.config.maxAllowedAnomalies) {
      if (status !== "OUTRIGHT_REJECTED") {
        status = "PARTIALLY_TRUSTED";
      }
      reason = `Anomaly count ${metrics.anomalyCount} exceeded max allowed ${this.config.maxAllowedAnomalies}.`;
    }

    return {
      overallTrustScore: metrics.overallTrustScore,
      safetyStatus: status,
      anomalyCount: metrics.anomalyCount,
      reason,
      metrics,
    };
  }
}
